Monika After Story/game/zz_backup.rpy

```diff
@@ -660,13 +662,23 @@ init -900 python:
         """
         try:
             p_savedir = os.path.normcase(renpy.config.savedir + "/")
-            p_name = "persistent"
-            numnum, numnum_del = __mas__backupAndDelete(p_savedir, p_name)
-            cal_name = "db.mcal"
-            __mas__backupAndDelete(p_savedir, cal_name, numnum=numnum)
+            is_pers_backup = persistent._mas_is_backup
+
+            try:
+                persistent._mas_is_backup = True
+                renpy.save_persistent()
+                numnum, numnum_del = __mas__backupAndDelete(p_savedir, "persistent")
```

Comment:
I assume without an except block this will just continue raising the exception? even so, its much easier to follow this if the exception is explicitly raised here.

Reply:
`try` has optional sub-blocks, `except`, `finally`, `else`. It makes sense to use only the ones you need. We don't care what it will raise or will it raise anything at all, we care about running the `finally` block no matter what happens. Linters would mark that as a "useless except as statement".

Reply:
ok but linters don't have to read the code later and maintain it, and we do care about the exception here because `numnum` is used after this block so if the exception isn't raised out of here then we'll get an irrelevant exception logged. (Even if the exception is raised out of here because python, I would prefer the explicit exception re-raise for readability.)

Reply:
To me it clutters the code, implying we're doing some work with exceptions, but we don't. We don't care about what's being raised, we care about the code in `finally` being run. If we get an exception, the code after this block wouldn't run.

If you meant whether or not that code suppresses the exception, no, it doesn't, we don't do anything to the exception raised (you could say we propagate it). If we crash there, the exception would be raised as usual, but the cleanup code in `finally` would be run. Which is the goal.

I'll add it for you, but I think it's a bad example of explicitness where it hurts readability.

Reply:
this is a recognizable pattern in different languages (minus except vs catch):
```python
try:
    # something
except Exception as e:
    raise e
finally:
    # something
```

`finally` is optional in other languages, but `except`/`catch` is not. I think python is in the wrong here for allowing `except` to be optional, and I would say the notation that was committed (blank except and raise) is also a bad python construct. Defining the exception and re-raising is a pattern that can be understood without having to know that python allows exceptions to propagate when the except block isn't there in a try-except pattern.

Reply:
And this is one of the reason I didn't want to add it. You got confused right away.

> the notation that was committed (blank except and raise) is also a bad python construct.

This actually the place where we *should* to use bare except, using `Exception` would limit it to only this type, which is incorrect - we want to run this code for all cases.

> Defining the exception and re-raising is a pattern that can be understood without having to know that python allows exceptions to propagate when the except block isn't there in a try-except pattern.

And for people that know about Python it makes the code look like it's doing some exception handling while it doesn't. Plus, makes you question why the bare except.

Reply:
>And this is one of the reason I didn't want to add it. You got confused right away.

Without except is worse - what happens if there is an exception? My first thought is that it ignores it since it isn't explicitly re-raised,

>This actually the place where we should use bare except (if you want to use except at all), using Exception would limit it to only this type, which is incorrect - we want to run this code for all cases.

The logging code in the outer exception catches from base Exception - so its following the same pattern. All exceptions should be using Exception as a base class, exceptions that aren't would have broken through the original try-except and probably all through renpy code anyway, so its not something that we'd have to handle.

> And for people that know about Python it makes the code look like it's doing some exception handling while it doesn't. Plus, makes you question why the bare except.

Those people will clearly see that it re-raises the exception. And don't do a bare except, do the `Exception as e` so there's no ambiguity. 

>I don't think trying to apply other langs syntax and logic to Python is always a good thing

It's applying conventional programming concepts for readability. Other pythonic notation like `<x> in <y>` or `a < x < b` is great for readability so we should do them. optional `except` is not because it just seems like the exception is ignored entirely.

Reply:
> so its following the same pattern. All exceptions should be using Exception as a base class

No, the block I added should catch *everything* (and by everything I mean all renpy exception and all base py exception.). That's the goal. What happens to the outside try/except isn't my concern. My concern is running the `finally` block I added.

Reply:
>My concern is running the finally block I added.

`finally` always runs after a try-except pattern, whether or not a bare except is used doesn't affect the execution of the `finally` block. I'm concerned with readability using a python construct that obscures behavior.

Reply:
> whether or not a bare except is used doesn't affect the execution of the finally block.

I believe it does. We have to use either implicit propagation of everything (no `except`), or explicit bare `except` with manual propagation of everything. I'm also not sure if that would clutter the trace stack with an extra exception or substitute it.
The `finally` part must be run no matter what, even if renpy decides to throw an error, or Python interpreter crashes.

Readability is important, but not when it hurts the code. There is a thing when the code is too explicit/has too many comments.

Reply:
>I believe it does. We have to use either implicit propagation of everything (no except), or explicit bare except with manual propagation of everything. I'm also not sure if that would clutter the trace stack with an extra exception or substitute it.

This is wrong, `finally` always runs after a try-except pattern in standard execution, with the only exceptions being outside influences like the computer dying or a sigkill ([py-doc](https://docs.python.org/3/reference/compound_stmts.html#finally-clause)).

>Readability is important, but not when it hurts the code. 

I don't think it hurts the code here. It's slightly more verbose for an obvious, no-prior-python-knowledge-required understanding of what happens to the exception. I would prefer the `except` over a comment here since its a conventional pattern.

Reply:
Example in the docs uses the syntax I used here before:
```py
    try:
        1/0
    finally:
        return 42
```
What happens if a renpy exception occurs and we don't catch it?

Reply:
Like I said earlier, I think python is wrong for allowing except to be optional. We don't need to follow "pythonic practices" if its bad for readability - like `try-else` for example. 

>What happens if a renpy exception occurs and we don't catch it?

if the exception is occurring within the python instance then `finally` should still run regardless. If its happening outside (like it kills the python execution), then it doesn't matter if there were an except here or not, `finally` wouldn't run.

Reply:
Ok, nvm, after looking at [wiki](https://en.wikipedia.org/wiki/Exception_handling_syntax), I see that there are other languages that don't require catch as well. Guess I gotta get used to it.